symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -211,43 +184,47 @@ instance
     type WitnessField (ArithmeticCircuit a p i) = WitnessF a (WitVar p i)
     witnessF (behead -> (c, o)) = o <&> \case
       ConstVar cv -> fromConstant cv
-      SysVar (InVar iv) -> at $ SysVar (InVar iv)
-      SysVar (NewVar nv) -> acWitness c ! nv
+      LinVar k (NewVar nv) b -> fromConstant k * (acWitness c ! nv) + fromConstant b
+      linInV -> at linInV
     fromCircuitF (behead -> (c, o)) f = uncurry (set #acOutput) (runState (f o) c)
 
 ----------------------------- MonadCircuit instance ----------------------------
 
 instance Finite a => Witness (Var a i) (WitnessF a (WitVar p i)) where
-  at (ConstVar cV) = fromConstant cV
-  at (SysVar sV)   = WitnessF (\x -> x (WSysVar sV))
+  at (ConstVar cV)   = fromConstant cV
+  at (LinVar k sV b) = fromConstant k * WitnessF (\x -> x (WSysVar sV)) + fromConstant b
 
 instance
   ( Arithmetic a, Binary a, Binary (Rep p), Binary (Rep i), Ord (Rep i)
   , o ~ U1) => MonadCircuit (Var a i) a (WitnessF a (WitVar p i)) (State (ArithmeticCircuit a p i o)) where
 
-    unconstrained wf = case runWitnessF wf $ const Nothing of
-      Just cV -> return (ConstVar cV)
-      Nothing -> do
-        let v = toVar @a wf
-        -- TODO: forbid reassignment of variables
-        zoom #acWitness $ modify (insert v wf)
-        return $ SysVar (NewVar v)
+    unconstrained wf =
+      case runWitnessF wf $ \case
+        WExVar _ -> More
+        WSysVar sV -> LinUVar one sV zero
+      of
+        ConstUVar c -> return $ ConstVar c
+        LinUVar k x b -> return $ LinVar k x b
+        More ->  do
+          let v = toVar @a wf
+          zoom #acWitness $ modify (insert v wf)
+          return $ toLinVar (NewVar v)
 
     constraint p =
       let evalConstVar = \case
-            SysVar sysV -> var sysV
+            LinVar k sysV b -> fromConstant k * (var sysV) + fromConstant b
             ConstVar cV -> fromConstant cV
           evalMaybe = \case
-            SysVar _ -> Nothing
+            LinVar {} -> Nothing
             ConstVar cV -> Just cV
       in case p evalMaybe of
-        Just c -> if c ==zero
+        Just c -> if c == zero
                     then return ()
                     else error "The constraint is non-zero"
         Nothing -> zoom #acSystem . modify $ insert (toVar @_ @p (p at)) (p evalConstVar)
 
-    rangeConstraint (SysVar v) upperBound =
-      zoom #acRange . modify $ insertWith S.union upperBound (S.singleton v)
+    rangeConstraint (LinVar k x b) upperBound =
+      zoom #acRange . modify $ insertWith S.union ((upperBound - b) // k) (S.singleton x)
```

Comment:
Oops! Looks like we completely wrecked the range constraints here
In finite fields, range constraint "y <= b" actually means "y \in {0, 1,...,b}" because there's no proper ordering actually
So "kx+b<=c" actually means "x \in {-b/k, (1-b)/k,..., (c-b)/k}"
Note that there's no guarantee at all that this is a contiguous interval from 0 to (c-b)/k
I think there's two ways to remedy this:
1. Give up and create a new variable `y:=kx+b` and range constrain it in case k and b mess things up (pretty much always except for either (k=1, b=0) or (k=-1, b=c))
2. Generalize range constraints to lookup constraints of arbitrary subsets of a field
Pinging @vlasin because he might find this interesting

Reply:
(@hovanja2011, thank you for finding this! I definitely missed this when thinking things out)